backend/hitas/views/housing_company.py

```diff
@@ -459,3 +461,41 @@ def get_detail_queryset(self):
     @staticmethod
     def get_filterset_class():
         return HousingCompanyFilterSet
+
+    @action(detail=True, methods=["PATCH"], url_path="batch-complete-apartments")
+    def batch_complete_apartments(self, request, **kwargs) -> Response:
+        housing_company_uuid = lookup_id_to_uuid(self.kwargs["uuid"], HousingCompany)
+
+        input_data = BatchCompleteApartmentsSerializer(data=request.data)
+        input_data.is_valid(raise_exception=True)
+
+        data = input_data.validated_data
+
+        query_set = Apartment.objects.filter(
+            building__real_estate__housing_company__uuid=housing_company_uuid,
+            completion_date=None,
+        )
+        if not data["all"]:
+            query_set = query_set.filter(
+                apartment_number__gte=data["range_start"],
+                apartment_number__lte=data["range_end"],
+            )
+
+        completed_apartment_count = query_set.update(completion_date=data["completion_date"])
+
+        result = {
+            "completed_apartment_count": completed_apartment_count,
+        }
+        return Response(data=result, status=status.HTTP_200_OK)
+
+
+class BatchCompleteApartmentsSerializer(serializers.Serializer):
+    completion_date = serializers.DateField()
+    range_start = serializers.IntegerField(min_value=0)
+    range_end = serializers.IntegerField(min_value=0)
+    all = serializers.BooleanField(default=False)
+
+    def validate(self, attrs: dict[str, Any]) -> dict[str, Any]:
+        if not attrs["all"] and attrs["range_start"] > attrs["range_end"]:
+            raise serializers.ValidationError("Range start must be less than range end.")
+        return attrs
```

Comment:
Would be good to also validate that both `range_start` and `range_end` exist, if `all` is False